streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsProducer.java

```diff
@@ -58,15 +60,14 @@
     private boolean transactionInitialized = false;
 
     public StreamsProducer(final Producer<byte[], byte[]> producer,
-                           final boolean eosEnabled,
-                           final LogContext logContext,
-                           final String applicationId) {
-        log = logContext.logger(getClass());
+                           final String applicationId,
+                           final LogContext logContext) {
+        log = Objects.requireNonNull(logContext.logger(getClass()), "logContext cannot be null");
         logPrefix = logContext.logPrefix().trim();
 
         this.producer = Objects.requireNonNull(producer, "producer cannot be null");
         this.applicationId = applicationId;
-        this.eosEnabled = eosEnabled;
+        this.eosEnabled = applicationId != null;
```

Comment:
We need the `application.id` only if EOS is enabled -- hence we simplify the constructor to 3 parameter and enable EOS is `application.id` is provided (this avoid the redundant `eosEnabled` parameter)